reps/2022-09-19-ray-on-spark.md

```diff
@@ -0,0 +1,258 @@
+and you could get very strange results with bad luck on the node sizing.
+
+
+#### Do Ray nodes have a 1:1 mapping with Spark tasks or with Spark workers?
+In each Spark task, we will launch one Ray worker node. All spark tasks are allocated the same
+number of resources. As a result, all Ray worker nodes will have homogeneous resources because Spark
+task resource configurations must be homogeneous.
+
+#### What is the recommended minimal resource allocation of a Ray node?
+
+Each Ray node should have at least 4 CPU cores and 10GB of available memory. This corresponds to
+a Spark task configuration of:
+
+- ``spark.task.cpus >= 4`` and
+- ``(SPARK_WORKER_MEMORY / TASK_SLOTS_PER_SPARK_WORKER) >= 10GB``
+
+
+#### On a shared spark cluster, shall we let each user launch individual ray cluster, or one user can create a shared ray cluster and other user can also use it ?
+I think we can provide 2 API to support both cases.
+Shared mode:
+```ray.spark.get_or_create_shared_cluster```
+and
+```ray.spark.init_private_cluster```
+
+And if one shared mode ray cluster already exists, new ray cluster is not allowed to be launched.
+
+
+#### How to select the port number used by Ray node ?
+Ray node requires listening on several port, a spark cluster might be shared by many users,
+each users might setup their own ray cluster concurrently, to avoid port conflicts,
+we can randomly select free port and start Ray node service,
+if failed we can retry on another free port.
+
+
+#### How much memory shall we allocate to Ray node service ( set via ray script --memory option) ?
+Spark does not provide explicit API for getting task allowed memory,
+SPARK_WORKER_MEMORY / SPARK_WORKER_CORES * RAY_NODE_NUM_CPUS
+
+
+#### How will we allocate memory for the Ray object store?
+We will calculate the memory reserved for the Ray object store in each Ray node as follows:
+
+1. Calculate the available space in the ``/dev/shm`` mount.
+2. Divide the ``/dev/shm`` available space by the number of Spark tasks (Ray nodes) that can run
+   concurrently on a single Spark worker node, obtaining an upper bound on the per-Ray-node object
+   store memory allocation.
+3. To provide a buffer and guard against accidental overutilization of mount space, multiply the
+   upper bound from (2) by ``0.8``.
+
+This allocation will ensure that we do not exhaust ``/dev/shm`` storage and accidentally spill to
+disk, which would hurt performance substantially.
+
+
+#### How to make ray respect spark GPU resource scheduling ?
+In spark task, we can get GPU IDs allocated to this task, so, when launching
+Ray node, besides specifying `--num-gpus` options, we need specify `CUDA_VISIBLE_DEVICES`
+environment so that we can restrict Ray node only uses the GPUs allocated to corresponding spark tasks.
+
+
+#### How to support all options of ray start
+Provide a ray_node_options argument (dict type).
+
+
+#### Where should the code live: ray repo or spark repo?
+Ray repo.
```

Comment:
Does it make sense to set up a separate repo, instead of using main ray repro?

Reply:
HI @scv119,
I work with Weichen and Corey. 
We definitely initially entertained the idea of creating a plugin provider to enable Ray support for Apache Spark. However, there are a few reasons why we think that might dramatically curtail the usage of such an integration. With the large volume of requests we've had in the OSS community for enabling a fully-supported integration, providing the community with a 3rd party (or 2nd party) plugin creates the impression of flippancy and increases the amount of effort that users would have to perform to get this much-asked-for functionality.
To whit, we'd like to see a true 1st party integration that is fully maintained by our Eng team. While having the integration as a plugin would certainly be simpler in implementation, we feel that the perception of support (most plugins aren't well maintained and that blanket perception generally extends to all, regardless of reality), as well as the additional feature parity / versioning complexity that would arise from maintaining a dependency between core Ray and a plugin would result in the solution in becoming effective shelf-ware. 
That being said, we understand and fully appreciate the gravity of such an integration living within the main Ray repo. Be assured that we would handle and assume ownership of the integration in the same way that we treat our own OSS projects with.  That being a production-level inclusion to our on-call staff members to ensure that issues, feature requests, and CI failures would be handled no differently than how we respond to Spark or MLflow addressable items. 

Hopefully that expresses our general thoughts on this. 
Kindest regards,
Ben

Reply:
Hi @scv119 , @ericl , @jjyao Any updates here?

Reply:
Q: Does ray team also helps maintaining projects like kuberay / xgboost_ray ? If these separate ray projects are maintained well, we could consider to make spark_ray as a separate project.

Reply:
> Q: Does ray team also helps maintaining projects like kuberay / xgboost_ray ? If these separate projects are maintained well, we could consider to make spark_ray as a separate project. A separate project has a benefit that we can build and locally run tests easier, otherwise we have to build the whole ray project, which is complex.

As @BenWilson2 , I would prefer to structure this as a contribution to the Ray repository to ensure full support and centralized communication / collaboration, especially since a lot of Spark users are asking for this.

@scv119 @ericl @jjyao What do we need to do to make a decision here? :)

Reply:
I think this would also be better for the Ray repository, especially if we want to test and maintain it as a first class integration.